fsspec/implementations/prefix.py

```diff
@@ -0,0 +1,148 @@
+        """
+        super().__init__(*args, **storage_options)
+        self.fs = fs
+
+        if not prefix:
+            raise ValueError("empty prefix is not a valid prefix")
+
+        self.prefix = prefix
+
+    def _add_fs_prefix(self, path):
+        if isinstance(path, str):
+            if self.prefix == self.fs.root_marker:
+                return f"{self.fs.root_marker}{path}"
+            return f"{self.prefix}{self.fs.sep}{path}"
+        return [self._add_fs_prefix(x) for x in path]
+
+    def _remove_fs_prefix(self, path):
+        if isinstance(path, str):
+            if self.prefix == self.fs.root_marker:
+                return path[len(self.prefix) :]
+            return path[len(self.prefix) + 1 :]
+        return [self._remove_fs_prefix(x) for x in path]
+
+    def mkdir(self, path, create_parents=True, **kwargs):
+        path = self._add_fs_prefix(path)
+        return self.fs.mkdir(path=path, create_parents=create_parents, **kwargs)
+
+    def makedirs(self, path, exist_ok=False):
+        path = self._add_fs_prefix(path)
+        return self.fs.makedirs(path=path, exist_ok=exist_ok)
+
+    def rmdir(self, path):
+        path = self._add_fs_prefix(path)
+        return self.fs.rmdir(path=path)
+
+    def ls(
+        self,
+        path,
+        detail=False,
+        **kwargs,
+    ):
+        path = self._add_fs_prefix(path)
+        ls_out = self.fs.ls(path=path, detail=detail, **kwargs)
+        if detail:
+            for out in ls_out:
+                out["name"] = self._remove_fs_prefix(out["name"])
+            return ls_out
+        return self._remove_fs_prefix(ls_out)
+
+    def glob(self, path, **kwargs):
+        path = self._add_fs_prefix(path)
+        glob_out = self.fs.glob(path=path, **kwargs)
+        return [self._remove_fs_prefix(x) for x in glob_out]
+
+    def info(self, path, **kwargs):
+        path = self._add_fs_prefix(path)
+        return self.fs.info(path=path, **kwargs)
+
+    def cp_file(self, path1, path2, **kwargs):
+        path1 = self._add_fs_prefix(path1)
+        path2 = self._add_fs_prefix(path2)
+        return self.fs.cp_file(path1, path2, **kwargs)
+
+    def get_file(self, path1, path2, callback=None, **kwargs):
+        path1 = self._add_fs_prefix(path1)
+        path2 = self._add_fs_prefix(path2)
+        return self.fs.get_file(path1, path2, callback=callback, **kwargs)
+
+    def put_file(self, path1, path2, callback=None, **kwargs):
+        path1 = self._add_fs_prefix(path1)
+        path2 = self._add_fs_prefix(path2)
+        return self.fs.put_file(path1, path2, callback=callback, **kwargs)
+
+    def mv_file(self, path1, path2, **kwargs):
+        path1 = self._add_fs_prefix(path1)
+        path2 = self._add_fs_prefix(path2)
+        return self.fs.mv_file(path1, path2, **kwargs)
+
+    def rm_file(self, path):
+        path = self._add_fs_prefix(path)
+        return self.fs.rm_file(path)
+
+    def rm(self, path, recursive=False, maxdepth=None):
```

Comment:
I think this and methods following should work based on the superclass implementation

Reply:
I agree, just wanted to avoid changing this too much right now. But might as well, alongside with `cat` too. Will do.

Reply:
Something I'm not sure about here: are we taking any care of sync/async? For example, [here](https://github.com/fsspec/filesystem_spec/blob/master/fsspec/implementations/reference.py#L213) we have both `_cat_file` and `cat_file` for the two cases.

Reply:
And [here](https://github.com/fsspec/filesystem_spec/blob/master/fsspec/implementations/reference.py#L246) we decide which to call. Maybe this pattern can be reused. In fact, there is a lot in common between prefix and reference.

Reply:
Sorry :|